lib/ood_core/job/adapters/htcondor.rb

```diff
@@ -0,0 +1,473 @@
+                    rescue Error => e
+                        raise Error, "Failed to hold job #{id}: #{e.message}"
+                    end
+
+                    # Release a job from hold using `condor_release`
+                    # @param id [#to_s] the id of the job to release
+                    # @raise [Error] if `condor_release` command exited unsuccessfully
+                    def release_job(id)
+                        id = id.to_s
+                        call("condor_release", id)
+                    rescue Error => e
+                        raise Error, "Failed to release job #{id}: #{e.message}"
+                    end
+
+                    # Retrieve job information using `condor_q`
+                    # @param id [#to_s] the id of the job
+                    # @param owner [String] the owner(s) of the job
+                    # @raise [Error] if `condor_q` command exited unsuccessfully
+                    # @return [Array<Hash>] list of details for jobs
+                    def get_jobs(id: "", owner: nil)
+                        args = []
+                        args.concat ["-constraint", "ClusterId == #{id}"] unless id.to_s.empty?
+                        args.concat ["-constraint", "Owner == \"#{owner}\""] unless owner.to_s.empty?
+                        args.concat ["-af", "ClusterId", "JobStatus", "Owner", "AcctGroup", "JobBatchName", "GlobalJobId", "CpusProvisioned", "GpusProvisioned", "QDate", "JobCurrentStartDate", "RemoteSysCpu", "RemoteUserCpu", "RemoteWallClockTime"]
+
+                        output = call("condor_q", *args)
+                        parse_condor_q_output(output)
+                    end
+
+                    # Retrieve slot information using `condor_status`
+                    # @param owner [String] the owner(s) of the slots
+                    # @raise [Error] if `condor_status` command exited unsuccessfully
+                    # @return [Array<Hash>] list of details for slots
+                    def get_slots
+                        args = ["-af", "Machine", "TotalSlotCPUs", "TotalSlotGPUs", "TotalSlotMemory", "CPUs", "GPUs", "Memory", "NumDynamicSlots"]
+                        args.concat ["-constraint", "\"DynamicSlot is undefined\""]
+
+                        output = call("condor_status", *args)
+                        parse_condor_status_output(output)
+                    end
+
+
+                    # Retrieve accounts using user_group_map on @submit_host
+                    # @return [Hash{String => Array<String>}] mapping of usernames to their groups
+                    def get_accounts
+                        raise Error, "user_group_map is not defined" if user_group_map.nil? || user_group_map.empty?
+                        
+                        # Retrieve accounts and cache them
+                        output = call("cat", user_group_map)
+                        accounts = {}
+                        output.each_line do |line|
```

Comment:
Seems like instead of calling `cat` you should just read the file. Casting that string to a `Pathname` you can do something like

```ruby
file = Pathname.new(user_group_map || "")
raise Error, "#{user_group_map} does not exist or cannot be read" unless file.exist? && file.readable?

file.readlines.each do |line|
  # all the other stuff, though you may need to line.chomp to get rid of the newline.
end
```

Reply:
Hmm... I guess we could require this to be _also_ present on the OOD host...
In our setup, this file is (just) deployed to the submission nodes (aka the one that we're ssh-ing to).
Reading this via SSH, ensures that the file read here corresponds with the file actually being used on the submission nodes